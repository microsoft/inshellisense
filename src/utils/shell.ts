// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import process from "node:process";
import find from "find-process";
import path from "node:path";
import which from "which";
import fs from "node:fs";
import url from "node:url";
import os from "node:os";
import fsAsync from "node:fs/promises";

export enum Shell {
  Bash = "bash",
  Powershell = "powershell",
  Pwsh = "pwsh",
  Zsh = "zsh",
  Fish = "fish",
  Cmd = "cmd",
}

export const supportedShells = [Shell.Bash, process.platform == "win32" ? Shell.Powershell : null, Shell.Pwsh, Shell.Zsh, Shell.Fish].filter(
  (shell) => shell != null,
) as Shell[];

export const userZdotdir = process.env?.ZDOTDIR ?? os.homedir() ?? `~`;
export const zdotdir = path.join(os.tmpdir(), `is-zsh`);

export const setupZshDotfiles = async () => {
  const shellFolderPath = path.join(path.dirname(url.fileURLToPath(import.meta.url)), "..", "..", "shell");
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-rc.zsh"), path.join(zdotdir, ".zshrc"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-profile.zsh"), path.join(zdotdir, ".zprofile"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-env.zsh"), path.join(zdotdir, ".zshenv"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-login.zsh"), path.join(zdotdir, ".zlogin"));
};

export const inferShell = async () => {
  try {
    const name = path.parse(process.env.SHELL ?? "").name;
    const shellName = supportedShells.find((shell) => name.includes(shell));
    if (shellName) return shellName;
  } catch {
    /* empty */
  }
  const processResult = (await find("pid", process.ppid)).at(0);
  const name = processResult?.name;
  return name != null ? supportedShells.find((shell) => name.includes(shell)) : undefined;
};

export const gitBashPath = async (): Promise<string> => {
  const gitBashPaths = await getGitBashPaths();
  for (const gitBashPath of gitBashPaths) {
    if (fs.existsSync(gitBashPath)) {
      return gitBashPath;
    }
  }
  throw new Error("unable to find a git bash executable installed");
};

const getGitBashPaths = async (): Promise<string[]> => {
  const gitDirs: Set<string> = new Set();

  const gitExePath = await which("git.exe", { nothrow: true });
  if (gitExePath) {
    const gitExeDir = path.dirname(gitExePath);
    gitDirs.add(path.resolve(gitExeDir, "../.."));
  }

  const addValid = <T>(set: Set<T>, value: T | undefined) => {
    if (value) set.add(value);
  };

  // Add common git install locations
  addValid(gitDirs, process.env["ProgramW6432"]);
  addValid(gitDirs, process.env["ProgramFiles"]);
  addValid(gitDirs, process.env["ProgramFiles(X86)"]);
  addValid(gitDirs, `${process.env["LocalAppData"]}\\Program`);

  const gitBashPaths: string[] = [];
  for (const gitDir of gitDirs) {
    gitBashPaths.push(
      `${gitDir}\\Git\\bin\\bash.exe`,
      `${gitDir}\\Git\\usr\\bin\\bash.exe`,
      `${gitDir}\\usr\\bin\\bash.exe`, // using Git for Windows SDK
    );
  }

  // Add special installs that don't follow the standard directory structure
  gitBashPaths.push(`${process.env["UserProfile"]}\\scoop\\apps\\git\\current\\bin\\bash.exe`);
  gitBashPaths.push(`${process.env["UserProfile"]}\\scoop\\apps\\git-with-openssh\\current\\bin\\bash.exe`);

  return gitBashPaths;
};
