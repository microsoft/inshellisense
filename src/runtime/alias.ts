// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import log from "../utils/log.js";
import { gitBashPath, Shell } from "../utils/shell.js";
import { CommandToken, parseCommand } from "./parser.js";
import { buildExecuteShellCommand } from "./utils.js";
import os from "node:os";

const loadedAliases: { [key: string]: CommandToken[] | undefined } = {};
const platform = os.platform();
const executeShellCommand = buildExecuteShellCommand(5_000);

const loadBashAliases = async () => {
  const shellTarget = platform == "win32" ? await gitBashPath() : Shell.Bash;
  const { stdout, stderr, status } = await executeShellCommand({ command: shellTarget, args: ["-i", "-c", "alias"], cwd: process.cwd() });
  if (status !== 0) {
    log.debug({ msg: "failed to load bash aliases", stderr, status });
    return;
  }

  return stdout
    .trim()
    .split("\n")
    .forEach((line) => {
      const [alias, ...commandSegments] = line.replace("alias ", "").replaceAll("'\\''", "'").split("=");
      loadedAliases[alias] = parseCommand(commandSegments.join("=").slice(1, -1) + " ");
    });
};

const loadZshAliases = async () => {
  const { stdout, stderr, status } = await executeShellCommand({ command: Shell.Zsh, args: ["-i", "-c", "alias"], cwd: process.cwd() });
  if (status !== 0) {
    log.debug({ msg: "failed to load zsh aliases", stderr, status });
    return;
  }

  return stdout
    .trim()
    .split("\n")
    .forEach((line) => {
      const [alias, ...commandSegments] = line.replaceAll("'\\''", "'").split("=");
      loadedAliases[alias] = parseCommand(commandSegments.join("=").slice(1, -1) + " ");
    });
};

export const loadAliases = async (shell: Shell) => {
  switch (shell) {
    case Shell.Bash:
      await loadBashAliases();
      break;
    case Shell.Zsh:
      await loadZshAliases();
      break;
  }
  return [];
};

export const aliasExpand = (command: CommandToken[]): CommandToken[] => {
  if (!command.at(0)?.complete) return command;

  const alias = loadedAliases[command.at(0)?.token ?? ""];
  if (alias) {
    log.debug({ msg: "expanding alias", alias, command: command.slice(1) });
    return [...alias, ...command.slice(1)];
  }
  return command;
};
