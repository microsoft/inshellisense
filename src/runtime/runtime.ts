// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import speclist, {
  diffVersionedCompletions as versionedSpeclist,
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
} from "@withfig/autocomplete/build/index.js";
import { parseCommand, CommandToken } from "./parser.js";
import { getArgDrivenRecommendation, getSubcommandDrivenRecommendation } from "./suggestion.js";
import { SuggestionBlob } from "./model.js";
import { buildExecuteShellCommand } from "./utils.js";

// eslint-disable-next-line @typescript-eslint/no-explicit-any -- recursive type, setting as any
const specSet: any = {};
(speclist as string[]).forEach((s) => {
  let activeSet = specSet;
  const specRoutes = s.split("/");
  specRoutes.forEach((route, idx) => {
    if (idx === specRoutes.length - 1) {
      const prefix = versionedSpeclist.includes(s) ? "/index.js" : `.js`;
      activeSet[route] = `@withfig/autocomplete/build/${s}${prefix}`;
    } else {
      activeSet[route] = activeSet[route] || {};
      activeSet = activeSet[route];
    }
  });
});

const loadedSpecs: { [key: string]: Fig.Spec } = {};

const loadSpec = async (cmd: CommandToken[]): Promise<Fig.Spec | undefined> => {
  const rootToken = cmd.at(0);
  if (!rootToken?.complete) {
    return;
  }

  if (loadedSpecs[rootToken.token]) {
    return loadedSpecs[rootToken.token];
  }
  if (specSet[rootToken.token]) {
    const spec = (await import(specSet[rootToken.token])).default;
    loadedSpecs[rootToken.token] = spec;
    return spec;
  }
};

// this load spec function should only be used for `loadSpec` on the fly as it is cacheless
const lazyLoadSpec = async (key: string): Promise<Fig.Spec | undefined> => {
  return (await import(`@withfig/autocomplete/build/${key}.js`)).default;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars -- will be implemented in below TODO
const lazyLoadSpecLocation = async (location: Fig.SpecLocation): Promise<Fig.Spec | undefined> => {
  return; //TODO: implement spec location loading
};

export const getSuggestions = async (cmd: string): Promise<SuggestionBlob | undefined> => {
  const activeCmd = parseCommand(cmd);
  const rootToken = activeCmd.at(0);
  if (activeCmd.length === 0 || !rootToken?.complete) {
    return;
  }

  const spec = await loadSpec(activeCmd);
  if (spec == null) return;
  const subcommand = getSubcommand(spec);
  if (subcommand == null) return;

  const result = await runSubcommand(activeCmd.slice(1), subcommand);
  if (result == null) return;
  const lastCommand = activeCmd.at(-1);
  const charactersToDrop = lastCommand?.complete ? 0 : lastCommand?.token.length ?? 0;
  return { ...result, charactersToDrop };
};

const getPersistentOptions = (persistentOptions: Fig.Option[], options?: Fig.Option[]) => {
  const persistentOptionNames = new Set(persistentOptions.map((o) => (typeof o.name === "string" ? [o.name] : o.name)).flat());
  return persistentOptions.concat(
    (options ?? []).filter(
      (o) => (typeof o.name == "string" ? !persistentOptionNames.has(o.name) : o.name.some((n) => !persistentOptionNames.has(n))) && o.isPersistent === true,
    ),
  );
};

// TODO: handle subcommands that are versioned
const getSubcommand = (spec?: Fig.Spec): Fig.Subcommand | undefined => {
  if (spec == null) return;
  if (typeof spec === "function") {
    const potentialSubcommand = spec();
    if (Object.prototype.hasOwnProperty.call(potentialSubcommand, "name")) {
      return potentialSubcommand as Fig.Subcommand;
    }
    return;
  }
  return spec;
};

const executeShellCommand = buildExecuteShellCommand(5000);

const genSubcommand = async (command: string, parentCommand: Fig.Subcommand): Promise<Fig.Subcommand | undefined> => {
  const subcommandIdx = parentCommand.subcommands?.findIndex((s) => s.name === command);
  if (subcommandIdx == null) return;
  const subcommand = parentCommand.subcommands?.at(subcommandIdx);
  if (subcommand == null) return;

  // this pulls in the spec from the load spec and overwrites the subcommand in the parent with the loaded spec.
  // then it returns the subcommand and clears the loadSpec field so that it doesn't get called again
  switch (typeof subcommand.loadSpec) {
    case "function": {
      const partSpec = await subcommand.loadSpec(command, executeShellCommand);
      if (partSpec instanceof Array) {
        const locationSpecs = (await Promise.all(partSpec.map((s) => lazyLoadSpecLocation(s)))).filter((s) => s != null) as Fig.Spec[];
        const subcommands = locationSpecs.map((s) => getSubcommand(s)).filter((s) => s != null) as Fig.Subcommand[];
        (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx] = {
          ...subcommand,
          ...(subcommands.find((s) => s?.name == command) ?? []),
          loadSpec: undefined,
        };
        return (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx];
      } else if (Object.prototype.hasOwnProperty.call(partSpec, "type")) {
        const locationSingleSpec = await lazyLoadSpecLocation(partSpec as Fig.SpecLocation);
        (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx] = {
          ...subcommand,
          ...(getSubcommand(locationSingleSpec) ?? []),
          loadSpec: undefined,
        };
        return (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx];
      } else {
        (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx] = { ...subcommand, ...partSpec, loadSpec: undefined };
        return (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx];
      }
    }
    case "string": {
      const spec = await lazyLoadSpec(subcommand.loadSpec as string);
      (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx] = { ...subcommand, ...(getSubcommand(spec) ?? []), loadSpec: undefined };
      return (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx];
    }
    case "object": {
      (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx] = { ...subcommand, ...(subcommand.loadSpec ?? {}), loadSpec: undefined };
      return (parentCommand.subcommands as Fig.Subcommand[])[subcommandIdx];
    }
    case "undefined": {
      return subcommand;
    }
  }
};

const getOption = (activeToken: CommandToken, options: Fig.Option[]): Fig.Option | undefined => {
  return options.find((o) => (typeof o.name === "string" ? o.name === activeToken.token : o.name.includes(activeToken.token)));
};

const getPersistentTokens = (tokens: CommandToken[]): CommandToken[] => {
  return tokens.filter((t) => t.isPersistent === true);
};

const getArgs = (args: Fig.SingleOrArray<Fig.Arg> | undefined): Fig.Arg[] => {
  return args instanceof Array ? args : args != null ? [args] : [];
};

const runOption = async (
  tokens: CommandToken[],
  option: Fig.Option,
  subcommand: Fig.Subcommand,
  persistentOptions: Fig.Option[],
  acceptedTokens: CommandToken[],
): Promise<SuggestionBlob | undefined> => {
  if (tokens.length === 0) {
    throw new Error("invalid state reached, option expected but no tokens found");
  }
  const activeToken = tokens[0];
  const isPersistent = persistentOptions.some((o) => (typeof o.name === "string" ? o.name === activeToken.token : o.name.includes(activeToken.token)));
  if ((option.args instanceof Array && option.args.length > 0) || option.args != null) {
    const args = option.args instanceof Array ? option.args : [option.args];
    return runArg(tokens.slice(1), args, subcommand, persistentOptions, acceptedTokens.concat(activeToken), true, false);
  }
  return runSubcommand(tokens.slice(1), subcommand, persistentOptions, acceptedTokens.concat({ ...activeToken, isPersistent }));
};

const runArg = async (
  tokens: CommandToken[],
  args: Fig.Arg[],
  subcommand: Fig.Subcommand,
  persistentOptions: Fig.Option[],
  acceptedTokens: CommandToken[],
  fromOption: boolean,
  fromVariadic: boolean,
): Promise<SuggestionBlob | undefined> => {
  if (args.length === 0) {
    return runSubcommand(tokens, subcommand, persistentOptions, acceptedTokens, true, !fromOption);
  } else if (tokens.length === 0) {
    return await getArgDrivenRecommendation(args, subcommand, persistentOptions, undefined, acceptedTokens, fromVariadic);
  } else if (!tokens.at(0)?.complete) {
    return await getArgDrivenRecommendation(args, subcommand, persistentOptions, tokens[0].token, acceptedTokens, fromVariadic);
  }

  const activeToken = tokens[0];
  if (args.every((a) => a.isOptional)) {
    if (activeToken.isOption) {
      const option = getOption(activeToken, persistentOptions.concat(subcommand.options ?? []));
      if (option != null) {
        return runOption(tokens, option, subcommand, persistentOptions, acceptedTokens);
      }
      return;
    }

    const nextSubcommand = await genSubcommand(activeToken.token, subcommand);
    if (nextSubcommand != null) {
      return runSubcommand(tokens.slice(1), nextSubcommand, persistentOptions, getPersistentTokens(acceptedTokens.concat(activeToken)));
    }
  }

  const activeArg = args[0];
  if (activeArg.isVariadic) {
    return runArg(tokens.slice(1), args, subcommand, persistentOptions, acceptedTokens.concat(activeToken), fromOption, true);
  } else if (activeArg.isCommand) {
    if (tokens.length <= 0) {
      return;
    }
    const spec = await loadSpec(tokens);
    if (spec == null) return;
    const subcommand = getSubcommand(spec);
    if (subcommand == null) return;
    return runSubcommand(tokens.slice(1), subcommand);
  }
  return runArg(tokens.slice(1), args.slice(1), subcommand, persistentOptions, acceptedTokens.concat(activeToken), fromOption, false);
};

const runSubcommand = async (
  tokens: CommandToken[],
  subcommand: Fig.Subcommand,
  persistentOptions: Fig.Option[] = [],
  acceptedTokens: CommandToken[] = [],
  argsDepleted = false,
  argsUsed = false,
): Promise<SuggestionBlob | undefined> => {
  if (tokens.length === 0) {
    return getSubcommandDrivenRecommendation(subcommand, persistentOptions, undefined, argsDepleted, argsUsed, acceptedTokens);
  } else if (!tokens.at(0)?.complete) {
    return getSubcommandDrivenRecommendation(subcommand, persistentOptions, tokens[0].token, argsDepleted, argsUsed, acceptedTokens);
  }

  const activeToken = tokens[0];
  const activeArgsLength = subcommand.args instanceof Array ? subcommand.args.length : 1;
  const allOptions = [...persistentOptions, ...(subcommand.options ?? [])];

  if (activeToken.isOption) {
    const option = getOption(activeToken, allOptions);
    if (option != null) {
      return runOption(tokens, option, subcommand, persistentOptions, acceptedTokens);
    }
    return;
  }

  const nextSubcommand = await genSubcommand(activeToken.token, subcommand);
  if (nextSubcommand != null) {
    return runSubcommand(
      tokens.slice(1),
      nextSubcommand,
      getPersistentOptions(persistentOptions, subcommand.options),
      getPersistentTokens(acceptedTokens.concat(activeToken)),
    );
  }

  if (activeArgsLength <= 0) {
    return; // not subcommand or option & no args exist
  }

  return runArg(tokens, getArgs(subcommand.args), subcommand, allOptions, acceptedTokens, false, false);
};
