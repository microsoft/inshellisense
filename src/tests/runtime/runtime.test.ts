// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { getSuggestions } from "../../runtime/runtime.js";
import { Shell } from "../../utils/shell.js";

const testData = [
  { name: "partialPrefixFilter", command: "git sta" },
  { name: "completePrefixFilter", command: "git stat" },
  { name: "emptySuggestions", command: "git to" },
  { name: "alreadyUsedSuggestion", command: "cd ~ " },
  { name: "alreadyUsedOption", command: "act --bind --b" },
  { name: "exclusiveOnOption", command: "ag --affinity --no" },
  { name: "providedSuggestion", command: "bw completion --shell " },
  { name: "fullyTypedSuggestion", command: "ls -W" },
  { name: "provideFolderSuggestion", command: "ls sr" },
  { name: "provideFileSuggestion", command: "ls READ" },
  { name: "optionsSuggestedAfterVariadicArg", command: "ls item -l", maxSuggestions: 3 },
  { name: "noOptionsSuggestedDuringVariadicArg", command: "ls -W ite" },
  { name: "providedArgDescription", command: "act completion bash -a " },
  { name: "completedOptionWithArg", command: "act completion bash -a 'actor' " },
  { name: "command", command: "sudo git sta" },
  { name: "nestedNonCommands", command: "az az ", skip: true }, // TODO: fix skipped test
  { name: "loadSpec", command: "aws acm add" },
  { name: "noArgsArgumentGiven", command: "gcc lab ", maxSuggestions: 3 },
  { name: "generatorUsingPartialInput", command: "dotnet add package Microsoft.Azure.WebJobs.Cor", maxSuggestions: 1 },
];

describe(`parseCommand`, () => {
  testData.forEach(({ command, name, skip, maxSuggestions }) => {
    if (skip) return;
    test(name, async () => {
      const suggestions = await getSuggestions(command, process.cwd(), Shell.Cmd);
      if (suggestions != null && suggestions.suggestions != null) {
        suggestions.suggestions = suggestions?.suggestions.slice(0, maxSuggestions);
      }
      expect(suggestions).toMatchSnapshot();
    });
  });
});
