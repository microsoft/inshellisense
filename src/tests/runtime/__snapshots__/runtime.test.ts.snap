// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCommand alreadyUsedOption 1`] = `
{
  "charactersToDrop": 3,
  "suggestions": [
    {
      "allNames": [
        "--bug-report",
      ],
      "description": "Display system information for bug report",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--bug-report",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand alreadyUsedSuggestion 1`] = `undefined`;

exports[`parseCommand command 1`] = `
{
  "charactersToDrop": 3,
  "suggestions": [
    {
      "allNames": [
        "stage",
      ],
      "description": "Add file contents to the staging area",
      "icon": "📦",
      "insertValue": undefined,
      "name": "stage",
      "priority": 50,
    },
    {
      "allNames": [
        "status",
      ],
      "description": "Show the working tree status",
      "icon": "📦",
      "insertValue": undefined,
      "name": "status",
      "priority": 50,
    },
    {
      "allNames": [
        "stash",
      ],
      "description": "Temporarily stores all the modified tracked files",
      "icon": "📦",
      "insertValue": undefined,
      "name": "stash",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand completePrefixFilter 1`] = `
{
  "charactersToDrop": 4,
  "suggestions": [
    {
      "allNames": [
        "status",
      ],
      "description": "Show the working tree status",
      "icon": "📦",
      "insertValue": undefined,
      "name": "status",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand completedOptionWithArg 1`] = `
{
  "charactersToDrop": 0,
  "suggestions": [
    {
      "allNames": [
        "--artifact-server-path",
      ],
      "description": "Defines the path where the artifact server stores uploads and retrieves downloads from. If not specified the artifact server will not start",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--artifact-server-path",
      "priority": 50,
    },
    {
      "allNames": [
        "--artifact-server-port",
      ],
      "description": "Defines the port where the artifact server listens (will only bind to localhost)",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--artifact-server-port",
      "priority": 50,
    },
    {
      "allNames": [
        "--container-architecture",
      ],
      "description": "Architecture which should be used to run containers, e.g.: linux/amd64. If not specified, will use host default architecture. Requires Docker server API Version 1.41+. Ignored on earlier Docker server platforms",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--container-architecture",
      "priority": 50,
    },
    {
      "allNames": [
        "--container-daemon-socket",
      ],
      "description": "Path to Docker daemon socket which will be mounted to containers",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--container-daemon-socket",
      "priority": 50,
    },
    {
      "allNames": [
        "--directory",
        "-C",
      ],
      "description": "Working directory",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--directory",
      "priority": 50,
    },
    {
      "allNames": [
        "--dryrun",
        "-n",
      ],
      "description": "Dryrun mode",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--dryrun",
      "priority": 50,
    },
    {
      "allNames": [
        "--env-file",
      ],
      "description": "Environment file to read and use as env in the containers",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--env-file",
      "priority": 50,
    },
    {
      "allNames": [
        "--github-instance",
      ],
      "description": "GitHub instance to use. Don't use this if you are not using GitHub Enterprise Server",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--github-instance",
      "priority": 50,
    },
    {
      "allNames": [
        "--insecure-secrets",
      ],
      "description": "NOT RECOMMENDED! Doesn't hide secrets while printing logs",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--insecure-secrets",
      "priority": 50,
    },
    {
      "allNames": [
        "--json",
      ],
      "description": "Output logs in json format",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--json",
      "priority": 50,
    },
    {
      "allNames": [
        "--no-recurse",
      ],
      "description": "Flag to disable running workflows from subdirectories of specified path in '--workflows'/'-W' flag",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--no-recurse",
      "priority": 50,
    },
    {
      "allNames": [
        "--no-skip-checkout",
      ],
      "description": "Do not skip actions/checkout",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--no-skip-checkout",
      "priority": 50,
    },
    {
      "allNames": [
        "--quiet",
        "-q",
      ],
      "description": "Disable logging of output from steps",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--quiet",
      "priority": 50,
    },
    {
      "allNames": [
        "--secret-file",
      ],
      "description": "File with list of secrets to read from (e.g. --secret-file .secrets)",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--secret-file",
      "priority": 50,
    },
    {
      "allNames": [
        "--verbose",
        "-v",
      ],
      "description": "Verbose output",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--verbose",
      "priority": 50,
    },
    {
      "allNames": [
        "--workflows",
        "-W",
      ],
      "description": "Path to workflow file(s)",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--workflows",
      "priority": 50,
    },
    {
      "allNames": [
        "--help",
        "-h",
      ],
      "description": "Display help",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--help",
      "priority": 50,
    },
    {
      "allNames": [
        "--no-descriptions",
      ],
      "description": "Disable completion descriptions",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--no-descriptions",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand emptySuggestions 1`] = `
{
  "charactersToDrop": 2,
  "suggestions": [],
}
`;

exports[`parseCommand exclusiveOnOption 1`] = `
{
  "charactersToDrop": 4,
  "suggestions": [
    {
      "allNames": [
        "--nobreak",
      ],
      "description": "Print a newline between matches in different files. Enabled by default",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nobreak",
      "priority": 50,
    },
    {
      "allNames": [
        "--nocolor",
      ],
      "description": "Don't print color codes in results",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nocolor",
      "priority": 50,
    },
    {
      "allNames": [
        "--nofilename",
      ],
      "description": "Don't print file names",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nofilename",
      "priority": 50,
    },
    {
      "allNames": [
        "--nofollow",
      ],
      "description": "Don't follow symlinks",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nofollow",
      "priority": 50,
    },
    {
      "allNames": [
        "--nogroup",
      ],
      "description": "Refrains from lumping matches in the same file together, and instead places the filename at the start of each match line",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nogroup",
      "priority": 50,
    },
    {
      "allNames": [
        "--noheading",
      ],
      "description": "Don't print filenames above matching contents",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--noheading",
      "priority": 50,
    },
    {
      "allNames": [
        "--nommap",
      ],
      "description": "Don't use of memory-mapped I/O. Defaults to true on platforms where mmap() is faster than read(). (All but macOS.)",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nommap",
      "priority": 50,
    },
    {
      "allNames": [
        "--nomultiline",
      ],
      "description": "Don't match regexes across newlines",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nomultiline",
      "priority": 50,
    },
    {
      "allNames": [
        "-n",
        "--norecurse",
      ],
      "description": "Don't recurse into directories",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--norecurse",
      "priority": 50,
    },
    {
      "allNames": [
        "--nonumbers",
      ],
      "description": "Don't print line numbers",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--nonumbers",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand fullyTypedSuggestion 1`] = `
{
  "charactersToDrop": 2,
  "suggestions": [
    {
      "allNames": [
        "-W",
      ],
      "description": "Display whiteouts when scanning directories.  (-S) flag)",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "-W",
      "priority": 50,
    },
    {
      "allNames": [
        "-w",
      ],
      "description": "Force raw printing of non-printable characters.  This is the default when output is not to a terminal",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "-w",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand generatorUsingPartialInput 1`] = `
{
  "charactersToDrop": 27,
  "suggestions": [
    {
      "allNames": [
        "Microsoft.Azure.WebJobs.Core",
      ],
      "description": "This library simplifies the task of adding background processing to your Microsoft Azure Web Sites. The SDK uses Microsoft Azure Storage, triggering a function in your program when items are added to Queues and Blobs. A dashboard provides rich monitoring and diagnostics for the programs that you write by using the SDK. For more information, please visit http://go.microsoft.com/fwlink/?LinkID=320971",
      "icon": "📀",
      "insertValue": "Microsoft.Azure.WebJobs.Core",
      "name": "Microsoft.Azure.WebJobs.Core",
      "priority": 60,
    },
  ],
}
`;

exports[`parseCommand loadSpec 1`] = `
{
  "charactersToDrop": 3,
  "suggestions": [
    {
      "allNames": [
        "add-tags-to-certificate",
      ],
      "description": "Adds one or more tags to an ACM certificate. Tags are labels that you can use to identify and organize your AWS resources. Each tag consists of a key and an optional value. You specify the certificate on input by its Amazon Resource Name (ARN). You specify the tag by using a key-value pair.  You can apply a tag to just one certificate if you want to identify a specific characteristic of that certificate, or you can apply the same tag to multiple certificates if you want to filter for a common relationship among those certificates. Similarly, you can apply the same tag to multiple resources if you want to specify a relationship among those resources. For example, you can add the same tag to an ACM certificate and an Elastic Load Balancing load balancer to indicate that they are both used by the same website. For more information, see Tagging ACM certificates.  To remove one or more tags, use the RemoveTagsFromCertificate action. To view all of the tags that have been applied to the certificate, use the ListTagsForCertificate action",
      "icon": "📦",
      "insertValue": undefined,
      "name": "add-tags-to-certificate",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand noArgsArgumentGiven 1`] = `
{
  "charactersToDrop": 0,
  "suggestions": [
    {
      "allNames": [
        "--analyzer-output",
      ],
      "description": "Static analyzer report output format",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--analyzer-output",
      "priority": 50,
    },
    {
      "allNames": [
        "--analyze",
      ],
      "description": "Run the static analyzer",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "--analyze",
      "priority": 50,
    },
    {
      "allNames": [
        "-arcmt-migrate-emit-errors",
      ],
      "description": "Emit ARC errors even if the migrator can fix them",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "-arcmt-migrate-emit-errors",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand noOptionsSuggestedDuringVariadicArg 1`] = `
{
  "charactersToDrop": 3,
  "suggestions": [],
}
`;

exports[`parseCommand optionsSuggestedAfterVariadicArg 1`] = `
{
  "argumentDescription": undefined,
  "charactersToDrop": 2,
  "suggestions": [
    {
      "allNames": [
        "package-lock.json",
      ],
      "description": undefined,
      "icon": "📄",
      "insertValue": undefined,
      "name": "package-lock.json",
      "priority": 55,
    },
    {
      "allNames": [
        "-L",
      ],
      "description": "Follow all symbolic links to final target and list the file or directory the link references rather than the link itself.  This option cancels the -P option",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "-L",
      "priority": 50,
    },
    {
      "allNames": [
        "-l",
      ],
      "description": "(The lowercase letter \`\`ell''.)  List in long format.  (See below.)  A total sum for all the file sizes is output on a line before the long listing",
      "icon": "🔗",
      "insertValue": undefined,
      "name": "-l",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand partialPrefixFilter 1`] = `
{
  "charactersToDrop": 3,
  "suggestions": [
    {
      "allNames": [
        "stage",
      ],
      "description": "Add file contents to the staging area",
      "icon": "📦",
      "insertValue": undefined,
      "name": "stage",
      "priority": 50,
    },
    {
      "allNames": [
        "status",
      ],
      "description": "Show the working tree status",
      "icon": "📦",
      "insertValue": undefined,
      "name": "status",
      "priority": 50,
    },
    {
      "allNames": [
        "stash",
      ],
      "description": "Temporarily stores all the modified tracked files",
      "icon": "📦",
      "insertValue": undefined,
      "name": "stash",
      "priority": 50,
    },
  ],
}
`;

exports[`parseCommand provideFileSuggestion 1`] = `
{
  "charactersToDrop": 4,
  "suggestions": [
    {
      "allNames": [
        "README.md",
      ],
      "description": undefined,
      "icon": "📄",
      "insertValue": undefined,
      "name": "README.md",
      "priority": 55,
    },
  ],
}
`;

exports[`parseCommand provideFolderSuggestion 1`] = `
{
  "charactersToDrop": 2,
  "suggestions": [
    {
      "allNames": [
        "src",
      ],
      "description": undefined,
      "icon": "📁",
      "insertValue": undefined,
      "name": "src",
      "priority": 55,
    },
  ],
}
`;

exports[`parseCommand providedArgDescription 1`] = `
{
  "argumentDescription": "actor",
  "charactersToDrop": 0,
  "suggestions": [],
}
`;

exports[`parseCommand providedSuggestion 1`] = `
{
  "argumentDescription": "Shell to generate completions for",
  "charactersToDrop": 0,
  "suggestions": [
    {
      "allNames": [
        "zsh",
      ],
      "description": undefined,
      "icon": "📀",
      "insertValue": undefined,
      "name": "zsh",
      "priority": 50,
    },
  ],
}
`;
